############################################################################################
# This script is used by LepMAP2pipeline.sh to convert a .csv file in chi square root table 
# format into a .linkage file.
#
# Input .csv file format:
# The chi square root table consists of one column listing SNP ids, and a column for 
# each sample. The data in the chi square root table is represented as either "H", "A", or
# something ("-" or "NA" for example, you can specify which you use in the 
# missingDataIndicator input parameter below) to represent missing or irrelevant data. The 
# most frequent genotype in an SNP is represented as H and the second most frequent type 
# is represented as A. 
#
#
# Information on .linkage file format (for use with Lep-MAP2 software):
#
# Data is tab-delimited.
#
# First 6 Columns contain "pedigree information":
# 1. Family ID (can be alphanumeric)
# 2. Individual ID (must be unique within family, can be alphanumeric)
# 3. Father ID (0 if father is not in family)
# 4. Mother ID (0 if mother is not in family)
# 5. Gender (0=unknown, 1=male, 2=female)
# 6. Affection status (0=unknown, 1=unaffected, 2=affected)
#
# Columns 7 and onward describe the phenotype data, separated by tabs. There 
# are four different types of phenotype data supported by the LINKAGE format 
# (Numbered Alleles, Binary Factors, Affection Status, Quantitative Traits), 
# and the Lep-MAP2 documentation uses Numbered Alleles.
#
# With Numbered Alleles, each genotype is represented as a pair of numbers 
# (eg: 1 2     2 2   1 1    1 2). Each number represents an allele, and 0 
# represents an unknown allele.
#
# For our purposes (input for the Lep-MAP2 software), we are setting:
# H = "1 2", A = "1 1", B = "2 2", NA = "0 0".
#
# Lep-MAP2 documentation: https://sourceforge.net/p/lepmap2/wiki/browse_pages/
#
# Official LINKAGE file format documentation available:
# http://www.jurgott.org/linkage/LinkagePC.html#__RefHeading__137_1806185151
# http://www.jurgott.org/linkage/LinkageUser.pdf
#
#
############################################################################################
# Input Parameters:
args <- commandArgs(trailingOnly = TRUE)

# path: The directory path of the folder that the input .csv file is in. This is also the 
# folder that the output .linkage file will be saved to. 
#path <- "/home/benrancourt/Downloads/r38-remove86genes/5272-remove86genes-try2/output/"
path <- args[1]

# filenameMinusExtension: The base name of the input .csv file, without the ".csv" postfix. This name will be
# used to read the .csv input file, but it will also be used as a basis for all the
# subsequent output files that will be generated by this pipeline.
#filenameMinusExtension <- "r38-5186genes-LG13separateFromLG12"
filenameMinusExtension <- args[2]

# familyName: This is probably the least important input parameter of this script. This is used as 
# a name for your input dataset when converting your input .csv file to .linkage format
# to be used as input for the LepMAP2 modules. The name you pick is important only to you.
#familyName <- "r38"
familyName <- args[3]

# missingDataIndicator: This parameter relates to the data in the original
# .csv data file that you supplied. Some of the data in your original input .csv will
# likely be missing, in which case it might be represented as NA or "-" or maybe even be
# empty, or something like that to indicate that it is missing.
# Indicate what your .csv file uses to represent missing data.
#missingDataIndicator <- "-"
missingDataIndicator <- args[4]






#########################################################################
# Execution Code:

options(stringsAsFactors = FALSE, warn = 1)

inputChiData <- read.csv(paste(path.expand(path),paste0(filenameMinusExtension,".csv"),sep="/"))

message("converting the chi square report to .linkage format")

if (!is.na(missingDataIndicator))
{
  if (missingDataIndicator == "NA")
  {
    missingDataIndicator <- NA
  }
}

reportLinkageGenotypes <- inputChiData[, -1] # remove the marker id column
message(paste0("ncols of original report: ",ncol(reportLinkageGenotypes)))
reportLinkageGenotypes <- cbind(parent1 = c("A"), parent2 = c("H"), reportLinkageGenotypes) # add two samples to use as parents
message(paste0("ncols of original report with two 'parent' cols added: ",ncol(reportLinkageGenotypes)))
message(paste0("nrows of original report: ",nrow(reportLinkageGenotypes)))
reportLinkageGenotypes <- t(reportLinkageGenotypes) # transpose the report (so it's columns are now rows)
message(paste0("nrows of transposed report (including 'parent' rows): ",nrow(reportLinkageGenotypes)))
message(paste0("ncols of transposed report: ",ncol(reportLinkageGenotypes)))


reportLinkageGenotypes[reportLinkageGenotypes=="H"] <- "1 2"
reportLinkageGenotypes[reportLinkageGenotypes=="A"] <- "1 1"
reportLinkageGenotypes[reportLinkageGenotypes=="B"] <- "2 2"
reportLinkageGenotypes[is.na(reportLinkageGenotypes)] <- "0 0"
reportLinkageGenotypes[reportLinkageGenotypes==missingDataIndicator] <- "0 0"
reportLinkageGenotypes[reportLinkageGenotypes=="-"] <- "0 0" # in case NA has been substituted with "-"
reportLinkageGenotypes[reportLinkageGenotypes==""] <- "0 0" # in case NA has been substituted with ""
reportLinkageGenotypes[reportLinkageGenotypes==" "] <- "0 0" # in case NA has been substituted with " "
reportLinkageGenotypes[reportLinkageGenotypes=="N/A"] <- "0 0" 
reportLinkageGenotypes[reportLinkageGenotypes=="na"] <- "0 0" 
reportLinkageGenotypes[reportLinkageGenotypes=="n/a"] <- "0 0" 


reportLinkage <- cbind(family = c(familyName), id = c(paste0("S",(1:nrow(reportLinkageGenotypes))-2)), fatherId = c("P1"), motherId = c("P2"), gender = c(0), affectionStatus = c(0), reportLinkageGenotypes)
reportLinkage[1,2:5] <- c("P1","0","0","1") # change id from S-1 to P1, no parents, male
reportLinkage[2,2:5] <- c("P2","0","0","2") # change id from S0 to P2, no parents, female

write.table(reportLinkage, file= paste(path.expand(path),paste0(filenameMinusExtension,".linkage"),sep="/"), append=FALSE, quote=FALSE, sep="\t", row.names=FALSE, col.names=FALSE)

message("report_gen part 2 complete")
